index.d.ts

```diff
@@ -0,0 +1,202 @@
+
+  interface IRouterMatcher<T> {
+    (path: PathParams, ...handlers: RouteHandler[]): T;
+    (path: PathParams, ...handlers: RequestHandlerParams[]): T;
+  }
+
+  interface IRouterHandler<T> {
+    (...handlers: RouteHandler[]): T;
+    (...handlers: RequestHandlerParams[]): T;
+  }
+
+  interface IRouter {
+    /**
+     * Map the given param placeholder `name`(s) to the given callback(s).
+     *
+     * Parameter mapping is used to provide pre-conditions to routes
+     * which use normalized placeholders. For example a _:user_id_ parameter
+     * could automatically load a user's information from the database without
+     * any additional code,
+     *
+     * The callback uses the same signature as middleware, the only differencing
+     * being that the value of the placeholder is passed, in this case the _id_
+     * of the user. Once the `next()` function is invoked, just like middleware
+     * it will continue on to execute the route, or subsequent parameter functions.
+     *
+     *      app.param('user_id', function(req, res, next, id){
+     *        User.find(id, function(err, user){
+     *          if (err) {
+     *            next(err);
+     *          } else if (user) {
+     *            req.user = user;
+     *            next();
+     *          } else {
+     *            next(new Error('failed to load user'));
+     *          }
+     *        });
+     *      });
+     */
+    param(name: string, handler: RequestParamHandler): this;
+
+    /**
+     * Alternatively, you can pass only a callback, in which case you have the opportunity to alter the app.param()
+     *
+     * @deprecated since version 4.11
+     */
+    param(callback: (name: string, matcher: RegExp) => RequestParamHandler): this;
+
+    /**
+     * Special-cased "all" method, applying the given route `path`,
+     * middleware, and callback to _every_ HTTP method.
+     */
+    all: IRouterMatcher<this>;
+
+    use: IRouterHandler<this> & IRouterMatcher<this>;
+
+    handle: (req: http.IncomingMessage, res: http.ServerResponse, cb: Callback) => void;
+
+    route(prefix: PathParams): IRoute;
+    // Stack of configured routes
+    stack: Layer[];
+
+    // Common HTTP methods
+    delete: IRouterMatcher<this>;
+    get: IRouterMatcher<this>;
+    head: IRouterMatcher<this>;
+    options: IRouterMatcher<this>;
+    patch: IRouterMatcher<this>;
+    post: IRouterMatcher<this>;
+    put: IRouterMatcher<this>;
+
+    // Exotic HTTP methods
+    acl: IRouterMatcher<this>;
+    bind: IRouterMatcher<this>;
+    checkout: IRouterMatcher<this>;
+    connect: IRouterMatcher<this>;
+    copy: IRouterMatcher<this>;
+    link: IRouterMatcher<this>;
+    lock: IRouterMatcher<this>;
+    "m-search": IRouterMatcher<this>;
+    merge: IRouterMatcher<this>;
+    mkactivity: IRouterMatcher<this>;
+    mkcalendar: IRouterMatcher<this>;
+    mkcol: IRouterMatcher<this>;
+    move: IRouterMatcher<this>;
+    notify: IRouterMatcher<this>;
+    pri: IRouterMatcher<this>;
```

Comment:
Is this actually a method on there?
```
$ node -pe 'require(".")().pri'
undefined
```

Reply:
I took that from `require('http').METHODS`. `v14.16.1` has it at least.

Reply:
Ah, gotcha. I was using Node.js 16.1. So that does bring up a good point: I guess this should be all the methods _possible_ in the various Node.js versions? I wonder what other methods got removed; probably should go through all the versions this module supports and union them together.

Reply:
I asked around in TS discord and they didn't recall a better way to handle it than what you proposed - adding collecting all of them from all versions.